src/lib/dfu.js

```diff
@@ -97,42 +118,74 @@ module.exports = {
 		});
 	},
 
-	findCompatibleDFU(showHelp = true) {
+	findCompatibleDFU({ showHelp = true, deviceId } = {}){
 		const { listDFUDevices, showDfuModeHelp } = module.exports;
 
 		return listDFUDevices()
 			.then((dfuDevices) => {
 				if (dfuDevices.length > 1) {
-					const question = {
-						type: 'list',
-						name: 'device',
-						message: 'Which device would you like to select?',
-						choices() {
-							return dfuDevices.map((d) => {
+					// Multiple devices in DFU mode
+					if (deviceId) {
+						// Look for the requested device
+						let matchingDevices = dfuDevices.filter((d) => {
+							return d.deviceId === deviceId;
+						});
+
+						// Too many or too few matches means something is wrong
+						if (matchingDevices.length === 0) {
+							return Promise.reject('No DFU device found matching the provided Device ID');
+						} else if (matchingDevices.length > 1) {
+							return Promise.reject('More than one DFU device found matching the provided Device ID');
+						}
+
+						return {
+							dfuId: matchingDevices[0].dfuId,
+							deviceId
+						};
+					} else {
+						// Ask which device to flash
+						const question = {
+							type: 'list',
+							name: 'device',
+							message: 'Which device would you like to select?',
+							choices() {
+								return dfuDevices.map((d) => {
+									return {
+										name: d.type,
+										value: d
+									};
+								});
+							}
+						};
+						return prompt([question])
+							.then((ans) => {
 								return {
-									name: d.type,
-									value: d.dfuId
+									dfuId: ans.device.dfuId,
+									deviceId: ans.device.deviceId
 								};
 							});
-						}
-					};
-					return prompt([question])
-						.then((ans) => {
-							const dfuId = ans.device;
-							module.exports.dfuId = dfuId;
```

Comment:
hm. ok. yeah.

![wheee!!](http://digitalspyuk.cdnds.net/16/26/640x352/gallery-1467024730-itcrowd-1.gif)

Reply:
(for clarity, i'm referring to our `exports` usage here not that this line changed 🤗)

Reply:
Ha! Yeah. Not sure I've ever seen `module.exports` used quite like that. I guess that's one way to force private members?

In any case, it makes refactoring DFU related code tricky.